File: 78_Subsets/Solution.py
class Solution:

    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        templist = []
        start = 0
        n = len(nums)
        self.backtracking(nums,start,templist,result,n)
        return result
    
    def backtracking(self,nums,start,templist,result,n):
        result.append(list(templist))
        for i in range(start,n):
            templist.append(nums[i])
            self.backtracking(nums,i+1,templist,result,n)
            templist.pop(-1)

s = Solution()
print(s.subsets([1,2,3]))

File: 78_Subsets/Solution.java
import java.util.*;
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> list = new ArrayList<>();
        Arrays.sort(nums);
        backtrack(list, new ArrayList<>(), nums, 0);
        return list;
    }
    
    private void backtrack(List<List<Integer>> list , List<Integer> tempList, int [] nums, int start){
        list.add(new ArrayList<>(tempList));
        for(int i = start; i < nums.length; i++){
            tempList.add(nums[i]);
            backtrack(list, tempList, nums, i + 1);
            tempList.remove(tempList.size() - 1);
        }
    }
    
    public static void main(String[] args) {
        Solution s  = new Solution();
        int[] nums = {1,2,3};
        for(List l : s.subsets(nums)){
            System.out.println(Arrays.toString(l.toArray()));
        }
    }
}